File: Array/union of two sorted arrays.cpp
//Brute
vector<int> sortedArray(vector<int>a, vector<int>b) {
  int n1=a.size();
  int n2=b.insert();
  set<int> st;  // contains unoque and sorted elements
  for(int i=0;i<n1;i++){   //itterate though the 1st array//
    st.insert(a[i]);
  }
  for(int i=0;i<n2;i++){   //iterate through the 2nd array//
    st.insert(b[i]);
  }
  vector<int> temp;       //store the elements form the set to the temp vector//
  for(auto it : st){
    temp.push_back(it);
  }
  return temp;
}

//optimal--->two pointer approach
vector<int>sortedArray(vector<int> a, vector<int> b){
  int n1=a.size();
  int n2=b.size();
  int i=0,j=0;
  vector<int> unionArr;
  while (i < n1 && j < n2) {
    if(a[i] <= b[j]) {
      if(unionArr.size()==0 || unionArr.back() != a[i]) {
        unionArr.push_back(a[i]);
      }
      i++;
    }
    else  {
      if(unionArr.size()==0 || unionArr.back() != b[j]) {
        unionArr.push_back(b[j]);
      }
      j++;
    }
  }
     while (i < n1 ) {
      if(unionArr.size()==0 || unionArr.back() != a[i]) {
        unionArr.push_back(a[i]);
      }
      i++;
    }
     while (j < n2 ) {
      if(unionArr.size()==0 || unionArr.back() != b[j]) {
        unionArr.push_back(b[j]);
      }
      j++;
    }  
}

File: Array/Union_of_two_sorted_arrays.java
✅ 1) Brute Force (using Set)

In Java, TreeSet automatically:
✔ keeps elements sorted
✔ stores unique values

import java.util.*;

class Solution {
    public static List<Integer> sortedArray(int[] a, int[] b) {
        int n1 = a.length;
        int n2 = b.length;

        Set<Integer> st = new TreeSet<>();  // sorted + unique

        for (int i = 0; i < n1; i++) {
            st.add(a[i]);
        }

        for (int i = 0; i < n2; i++) {
            st.add(b[i]);
        }

        // move set elements to list
        List<Integer> temp = new ArrayList<>(st);
        return temp;
    }
}
Time Complexity: O((n1+n2) log(n1+n2))

✅ 2) Optimal – Two Pointer Approach

Same logic as your C++ code.

import java.util.*;

class Solution {
    public static List<Integer> sortedArray(int[] a, int[] b) {
        int n1 = a.length;
        int n2 = b.length;

        int i = 0, j = 0;
        List<Integer> unionArr = new ArrayList<>();

        while (i < n1 && j < n2) {
            if (a[i] <= b[j]) {
                if (unionArr.size() == 0 || unionArr.get(unionArr.size() - 1) != a[i]) {
                    unionArr.add(a[i]);
                }
                i++;
            } else {
                if (unionArr.size() == 0 || unionArr.get(unionArr.size() - 1) != b[j]) {
                    unionArr.add(b[j]);
                }
                j++;
            }
        }

        while (i < n1) {
            if (unionArr.size() == 0 || unionArr.get(unionArr.size() - 1) != a[i]) {
                unionArr.add(a[i]);
            }
            i++;
        }

        while (j < n2) {
            if (unionArr.size() == 0 || unionArr.get(unionArr.size() - 1) != b[j]) {
                unionArr.add(b[j]);
            }
            j++;
        }

        return unionArr;
    }
}


Time Complexity: O(n1 + n2)
👉 Best for interviews.
